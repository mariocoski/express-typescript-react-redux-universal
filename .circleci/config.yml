version: 2
jobs:
  build:
    working_directory: ~/express-typescript-react-redux-universal/api
    # The primary container is an instance of the first list image listed. Your build commands run in this container.
    docker:
      - image: circleci/node:4.8.2

    steps:
      - checkout
      - run:
          name: Update npm
          command: 'sudo npm install -g npm@latest'
      - run:
          name: Install npm
          command: sudo npm install
      - run:
          name: tests
          command: sudo npm run test
  test:
    docker:
      - image: circleci/node:4.8.2
    steps:
      - checkout
      - run:
          name: Test
          command: yarn test

workflows:
  version: 2
  build_and_test:
    jobs:
      - build
      - test:
          requires:
            - build
          filters:
            branches:
              only: develop