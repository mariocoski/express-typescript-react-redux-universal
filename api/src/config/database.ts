const env = require('../utils').env; 

module.exports = {
  development: {
    username: 'root',
    password: 'root',
    database: 'database_dev',
    host: '127.0.0.1',
    dialect: 'mysql'
  },
  test: {
    username: env('TEST_DB_USERNAME'),
    password: env('TEST_DB_PASSWORD'),
    database: env('TEST_DB_NAME'),
    dialect: 'sqlite'
  },
  production: {
    username: process.env.PROD_DB_USERNAME,
    password: process.env.PROD_DB_PASSWORD,
    database: process.env.PROD_DB_NAME,
    host: process.env.PROD_DB_HOSTNAME,
    dialect: 'mysql'
  }
}