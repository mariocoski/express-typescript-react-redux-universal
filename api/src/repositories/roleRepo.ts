const db = require('../models');

export async function associateRole(user_id: number, role_id: number) {
  return db.users_roles.create({user_id, role_id});
}

export default {
  associateRole
}
