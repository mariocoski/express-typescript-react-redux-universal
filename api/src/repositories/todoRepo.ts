const db = require('../models');
import * as _ from 'lodash';

export async function createTodo(values: Object, settings: Object = {}) {
  const options = { fields: 
    ['user_id','title', 'description', 'completed_at', 'deleted_at', 'created_at', 'updated_at'],
    ...settings 
  };
  return db.Todo.create(values, options);
}

export async function getTodosForUserId(user_id: string) {
  return db.Todo.findAll({ where: { user_id, deleted_at: null } });
}

export async function getTodoById(todo_id: string){
  return db.Todo.findOne({ where: { deleted_at: null } });
}

export async function updateTodoById(todoId: number, data: any) {
  const fillable = ['title','description', 'completed_at'];
  
  const validData: any = _.pick(data, fillable);

  return db.Todo.update(validData, {
    where: {
      id: todoId
    }
  });
}

export async function deleteTodoById(todoId: number) {
  return db.Todo.destroy({
    where: {
      id: todoId
    }
  });
}