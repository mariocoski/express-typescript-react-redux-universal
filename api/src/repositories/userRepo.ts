const db = require('../models');


export async function findUserById(id: string) {
  return db.User.findById(id);
}

export async function findUserByEmail(email: string) {
  return db.User.findOne({ where: { email } });
}

export async function findUserByToken(token: string){
  return db.User.findOne({ where: { password_reset_token: token } });
}

export async function createUser(values: Object, settings: Object = {}) {
  const options = { fields: 
    ['email','password', 'first_name', 'last_name', 'bio','password_reset_token', 'password_reset_token_expired_at'],
    ...settings 
  };
  return db.User.create(values, options);
}

export default {
  findUserByToken,
  findUserByEmail,
  createUser
}
