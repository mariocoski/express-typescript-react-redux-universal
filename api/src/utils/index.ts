import {ROLE_USER, ROLE_PROFESSIONAL, ROLE_ADMIN, ROLE_SUPERADMIN} from '../constants/roles';
import {Request, Response} from 'express';
import {User} from '../models/user';

export function resolvePort(val: any) {
    const port = parseInt(val, 10);
  
    if (isNaN(port)) {
      return val;
    }
  
    if (port >= 0) {
      return port;
    }
    
    return 3000;
}



export const setUserInfo = (request) => {
  const getUserInfo = {
    _id: request._id,
    firstName: request.profile.firstName,
    lastName: request.profile.lastName,
    email: request.email,
    role: request.role
  };

  return getUserInfo;
};

export const getRole = (checkRole) => {
  let role;

  switch (checkRole) {
    case ROLE_SUPERADMIN: role = 4; break;
    case ROLE_ADMIN: role = 3; break;
    case ROLE_PROFESSIONAL: role = 2; break;
    case ROLE_USER: role = 1; break;
    default: role = 1;
  }

  return role;
};

export function isAuthorized(role: String){
  return function(req: Request,res: Response,next:Function){
    const user = req.user;
   
    User.findById(user._id, (err, foundUser:any)=>{
  
      if(err) {
        res.status(422).json({error: 'No user found'});
        return next(err);
      }
      if(getRole(foundUser.role) >= getRole(role)){
        return next();
      }
      res.status(401).json({error: "You are not authorized to see the content"});
      return next(err);
    });
   
    
  }
}


export function env(key): string {
  const myVal: string | undefined = process.env[key];
  if(myVal === undefined){
      throw new Error(`No ENV ${key} not found`);
  }
  return myVal;
}