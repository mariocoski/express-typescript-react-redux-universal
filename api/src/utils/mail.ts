
export async function sendEmail(data: Mailgun.messages.SendData,mailgun: Mailgun.Mailgun){
  return mailgun.messages().send(data);
}
