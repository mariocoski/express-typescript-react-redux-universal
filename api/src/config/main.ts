import { env } from '../utils';
export default {
  mailgun_domain: env('MAILGUN_DOMAIN'),
  mailgun_api_key: env('MAILGUN_API_KEY')
};