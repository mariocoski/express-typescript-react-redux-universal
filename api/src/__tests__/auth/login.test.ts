require('dotenv').config();
import {EMAIL_IS_REQUIRED, EMAIL_IS_INVALID, 
        PASSWORD_IS_REQUIRED, INVALID_CREDENTIALS} from '../../constants/errors';
const db = require('../../models');
import {seedDb} from '../../utils';
import {expectError} from '../helpers';
import config from '../../config/main';
import {findUserByEmail} from '../../repositories/userRepo';

describe('LOGIN', () => {
  const request = require('supertest');
  let app: any;

  beforeEach(async() => {
    await db.sequelize.sync({force:true});
    await seedDb(db.sequelize.queryInterface);
    app = require('../../server');
  });

  afterEach(async() => {
    app.close();
  });

  it('should fail to log in without input', async () => {
    expect.assertions(2);
    const response = await request(app).post('/api/v1/auth/login');
    expectError(response,EMAIL_IS_REQUIRED);
  });

  it('should fail to log in when email is invalid', async () => {
    const response = await request(app).post('/api/v1/auth/login')
                                       .type('form')
                                       .send( { email: 'invalid@email' });
    expectError(response,EMAIL_IS_INVALID);
  });

  it('should fail to log in without password', async () => {
    const response = await request(app).post('/api/v1/auth/login')
                                       .type('form')
                                       .send( { email: config.mailgun_test_recipient });
    expectError(response,PASSWORD_IS_REQUIRED);
  });

  it('should fail to log in when user does not exist', async () => {
    const response = await request(app).post('/api/v1/auth/login')
                                       .type('form')
                                       .send( { 
                                         email: 'user_does_not_exist@email.com',
                                         password: 'password'
                                       });
    expectError(response,INVALID_CREDENTIALS);
  });

  it('should fail to log in when user provided wrong password', async () => {
    await db.User.create({email:config.mailgun_test_recipient, password: 'password'});
    const response = await request(app).post('/api/v1/auth/login')
                                       .type('form')
                                       .send( { 
                                         email: config.mailgun_test_recipient,
                                         password: 'wrong_password'
                                       });
    expectError(response,INVALID_CREDENTIALS);
  });

  it('should login user with valid credentials', async () => {

    const validCredentials = {
       email:config.mailgun_test_recipient, 
       password: 'password', 
       password_reset_token: 'not null - clear up when user know password',
       password_reset_token_expired_at: Date.now()
    };
    
    const user = await db.User.create(validCredentials);
    const response = await request(app).post('/api/v1/auth/login')
                                       .type('form')
                                       .send({ 
                                         email:config.mailgun_test_recipient, 
                                         password: 'password'
                                       });
    const data = JSON.parse(response.text);
    const loggedInUser = await findUserByEmail(config.mailgun_test_recipient);
 
    expect(response.statusCode).toBe(200);
    expect(data.token).toMatch(/JWT/);
    expect(data.user).toMatchSnapshot();
    expect(loggedInUser.password_reset_token).toBeNull();
    expect(loggedInUser.password_reset_token_expired_at).toBeNull();
  });

});