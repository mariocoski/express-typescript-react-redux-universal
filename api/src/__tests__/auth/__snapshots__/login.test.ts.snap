// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LOGIN should login user with valid credentials 1`] = `
Object {
  "_id": 4,
  "email": "mariuszrajczakowski@gmail.com",
  "first_name": null,
  "last_name": null,
  "roles": Array [
    "user",
  ],
}
`;
